Reland: Plumb dart timeline events to mojo:tracing.

ppiet opened the pull request:
This updates and relands https://github.com/flutter/engine/commit/ef589f99bca883bf27a2596f75a531c7fd517774 which was reverted due to a build-breaking conflict with a recent mojo roll. Please find the original description below.

This patch:

 - makes dart traces (coming from the VM and the timeline API)
   available to mojo:tracing
 - adds --vm-complete-timeline parameter, which makes the Dart VM enable all
   categories of timeline traces from the beginning, allowing to do startup
   tracing

This is based on the equivalent support in dart content handler.

Fixes https://github.com/flutter/flutter/issues/3162.

ppiet commented:
Note that one can't currently test if this works as mojo:tracing seems to be broken, see https://github.com/domokit/mojo/issues/775 . Please consider this on hold - I will confirm that it works and update this thread when mojo tracing is fixed and rolled into engine.

cc @tonygentilcore

eseidelGoogle commented:
Since @abarth is out, I would like @johnmccutchan to explicitly sign off on this patch and merge it.

johnmccutchan commented:
@ppiet what is the delta between this version and the one that broke the build?

ppiet commented:
@johnmccutchan please find the diff attached. This corresponds to the changes to dart_tracing.{h,cc} made in mojo in  https://codereview.chromium.org/1918333003 and https://codereview.chromium.org/1928183002 .

[diff.txt](https://github.com/flutter/engine/files/261891/diff.txt)

johnmccutchan commented on sky/engine/core/script/dart_init.cc:
```diff
@@ -114,6 +114,10 @@ static const char* kDartTraceStartupArgs[]{
     "--timeline_recorder=endless",
 };
 
+static const char* kVmCompleteTimelineArgs[]{
```
Why is this necessary? The kDartTraceStartupArgs should give you what you need.

ppiet replied:
Nice, done. ("--trace-startup" used not to set timeline recorder to endless back at the time of the first patchset)

johnmccutchan commented on sky/shell/platform/mojo/dart_tracing.cc:
```diff
@@ -0,0 +1,172 @@
+  if (categories == mojo::String("Dart")) {
+    Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_DART);
+  } else {
+    // TODO(johnmccutchan): Respect |categories|.
```
s/johnmccutchan/ppiet.

ppiet replied:
Done.

johnmccutchan commented on sky/shell/platform/mojo/dart_tracing.cc:
```diff
@@ -0,0 +1,172 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
This file is basically the same as the one in Mojo, right? Is there a way to share the code?

ppiet replied:
The purpose is the same. We can put this under mojo/common - but since this depends on dart/, we would need to roll dart into engine in lockstep with mojo. Or maybe dart_tools_api.h is stable enough that this will not be a problem in practice?

johnmccutchan commented on sky/shell/platform/mojo/main_mojo.cc:
```diff
@@ -16,12 +16,14 @@
 namespace {
 
 const char kEnableCheckedMode[] = "--enable-checked-mode";
+const char kVmCompleteTimeline[] = "--vm-complete-timeline";
```
Flutter already has --trace-startup which will give you the complete timeline (minus events you don't care about).

ppiet replied:
Yay, done.

abarth commented on sky/shell/platform/mojo/dart_tracing.cc:
```diff
@@ -0,0 +1,172 @@
+  // The data from the VM is null-terminated content of a JSON array without its
+  // enclosing square brackets. Remove the trailing "\0" and add the enclosing
+  // brackets to make JSON parser happy.
+  std::string json_data = "[" + std::string(data, length -1) + "]";
```
This is very inefficient.

ppiet replied:
Switched to std::stringstream.

abarth commented on sky/shell/platform/mojo/dart_tracing.h:
```diff
@@ -0,0 +1,61 @@
+
+namespace dart {
+
+class DartTimelineController {
```
One class per file please.

ppiet replied:
Done. Also made all these static methods just global functions in anonymous namespace and dropped DartTimelineController.

abarth commented on sky/shell/platform/mojo/dart_tracing.h:
```diff
@@ -0,0 +1,61 @@
+#include "mojo/public/cpp/bindings/interface_request.h"
+#include "mojo/services/tracing/interfaces/tracing.mojom.h"
+
+namespace dart {
```
This namespace isn't correct.  Everything else in this directory is in the `sky::shell` namespace.

ppiet replied:
Done.

ppiet commented:
Thanks, @johnmccutchan and @abarth ! I addressed your comments, but haven't tested the newest patchset yet - waiting for the mojo fix for tracing to roll into engine and elsewhere. You might want to hold off until then, I will update this thread when this is verified.

abarth commented:
I'm working on a mojo roll in https://github.com/flutter/engine/pull/2717.  Just waiting for some iOS issues to stablize before landing.

Hixie commented:
@ppiet did you have a chance to look at this since the mojo roll?

ppiet commented:
@Hixie no. Mojo with the tracing fix rolled into Flutter (thanks!), but not yet into the consumer of this feature.

Hixie commented:
@ppiet Ok. I'm going to close this PR for now so that it isn't on our review-needed radar, please don't hesitate to reopen it when you're ready!
